File: apps/backend/internal/types/student.go
package types

import "github.com/jackc/pgx/v5/pgtype"

type CreateStudentRequest struct {
	FullName      string   `json:"fullName" binding:"required,notblank,min=2,max=70,alphaspace"`
	DobAD         string   `json:"dobAD" binding:"required,notblank,date_format"`
	DobBS         string   `json:"dobBS" binding:"required,notblank,date_format"`
	Phone         string   `json:"phone" binding:"required,notblank,nepal_phone"`
	Address       string   `json:"address" binding:"required,notblank,max=70"`
	PhotoUrl      string   `json:"photoUrl" binding:"omitempty,url"`
	Source        string   `json:"source" binding:"required,oneof=facebook instagram tiktok referral inperson"`
	GuardianName  string   `json:"guardianName" binding:"required,notblank,min=2,max=70"`
	GuardianPhone string   `json:"guardianPhone" binding:"required,notblank,nepal_phone"`
	Gender        string   `json:"gender" binding:"required,oneof=male female other"`
	Shift         string   `json:"shift" binding:"required,oneof=morning day evening"`
	ShiftTime     string   `json:"shiftTime" binding:"required,notblank,min=1,max=100"`
	Courses       []string `json:"courses" binding:"required,min=1,dive,uuid"`
}

type CreateStudentInquiryRequest struct {
	FullName string `json:"fullName" binding:"required,notblank,alphaspace"`
	Phone    string `json:"phone" binding:"required,notblank,nepal_phone"`
	Message  string `json:"message" binding:"required,notblank,min=5,max=200"`
	Email    string `json:"email" binding:"required,email"`
	Source   string `json:"source" binding:"required,oneof= facebook instagram tiktok referral inPerson"`
}

type Student struct {
	ID            pgtype.UUID `json:"id"`
	FullName      string      `json:"fullName"`
	DOB           pgtype.Date `json:"dob"`
	Phone         string      `json:"phone"`
	Email         pgtype.Text `json:"email"`
	Address       string      `json:"address"`
	PhotoUrl      pgtype.Text `json:"photoUrl"`
	Source        string      `json:"source"`
	GuardianName  string      `json:"guardianName"`
	GuardianPhone string      `json:"guardianPhone"`
	Gender        string      `json:"gender"`
	FiscalYear    string      `json:"fiscalYear"`
	ReferenceNo   string      `json:"referenceNo"`
	SerialNo      int         `json:"serialNo"`
	Notes         pgtype.Text `json:"notes"`
	Status        string      `json:"status"`
}
type ListStudent struct {
	ID          pgtype.UUID `json:"id"`
	FullName    string      `json:"fullName"`
	Phone       string      `json:"phone"`
	ReferenceNo string      `json:"referenceNo"`
	Status      string      `json:"status"`
	Courses     []string    `json:"courses"`
	Batch       string      `json:"batch"`
	Shift       string      `json:"shift"`
}

type UpdateStudentStatusRequest struct {
	Status          string `json:"status" binding:"required,oneof=pending active completed rejected"`
	RejectionReason string `json:"rejectionReason" binding:"required_if=Status rejected,omitempty,min=1,max=500"`
}

File: packages/types/src/student.ts
import z from "zod";

export const inquiryFormSchema = z.object({
  fullName: z
    .string()
    .min(2, "Full name must be at least 2 characters")
    .max(70, "Full name must be 70 characters or less")
    .regex(/^[A-Za-z ]+$/, "Full name can only contain letters and spaces")
    .trim(),

  phone: z
    .string()
    .trim()
    .regex(/^(98|97)\d{8}$/, "Enter a valid Nepali phone number"),

  message: z
    .string()
    .trim()
    .min(10, "Message must be at least 10 characters")
    .max(200, "Message must be 200 characters or less"),

  email: z.email("Enter a valid email address").trim(),

  source: z.enum(["facebook", "instagram", "tiktok", "referral", "inPerson"], {
    error: "Please select how you heard about us",
  }),
});

export type InquiryForm = z.infer<typeof inquiryFormSchema>;

export const createStudentAdmissionRequest = z.object({
  fullName: z
    .string()
    .trim()
    .min(1, "Full name is required")
    .min(5, "Full name must be at least 5 characters")
    .max(50, "Full name must be less than 70 characters")
    .regex(/^[A-Za-z ]+$/, "Full name can only contain letters and spaces"),

  phone: z
    .string()
    .trim()
    .min(1, "Phone number is required")
    .regex(/^(98|97)\d{8}$/, "Enter a valid Nepali phone number"),

  source: z.enum(["facebook", "instagram", "tiktok", "referral", "inPerson"], {
    error: "Please select how you heard about us",
  }),

  email: z
    .email("Enter a valid email address")
    .trim()
    .min(1, "Email is required"),

  dob: z
    .string()
    .min(1, "Date of birth is required")
    .regex(/^\d{4}-\d{2}-\d{2}$/, "Use format YYYY-MM-DD"),

  gender: z.enum(["male", "female", "other"], {
    error: "Please select a gender",
  }),

  guardianName: z
    .string()
    .trim()
    .min(1, "Guardian name is required")
    .min(5, "Guardian name must be at least 5 characters")
    .max(70, "Guardian name must be less than 70 characters")
    .regex(/^[A-Za-z ]+$/, "Guardian name can only contain letters and spaces"),

  guardianPhone: z
    .string()
    .trim()
    .min(1, "Guardian phone number is required")
    .regex(/^(98|97)\d{8}$/, "Enter a valid Nepali phone number"),

  courses: z.array(z.uuid()).min(1, "Please select at least one course"),

  address: z
    .string()
    .trim()
    .min(1, "Address is required")
    .min(5, "Address is too short"),

  claimedAmount: z
    .number({ error: "Amount must be a number" })
    .min(0, "Amount cannot be negative")
    .nullable(),

  photoUrl: z.url("Invalid photo URL").min(1, "Photo is required"),
});

export type CreateStudentAdmission = z.infer<
  typeof createStudentAdmissionRequest
>;
